src/generate.rs

```diff
@@ -0,0 +1,211 @@
+    /// can be omitted from the expression.
+    ///
+    /// For example:
+    ///
+    /// `|json_parser` prints a `json_parser` transform.
+    ///
+    /// `||file,http` prints a `file` and `http` sink.
+    ///
+    /// `stdin||http` prints a `stdin` source and an `http` sink.
+    ///
+    /// Vector makes a best attempt at constructing a sensible topology. The
+    /// first transform generated will consume from all sources and subsequent
+    /// transforms will consume from their predecessor. All sinks will consume
+    /// from the last transform or, if none are specified, from all sources. It
+    /// is then up to you to restructure the `inputs` of each component to build
+    /// the topology you need.
+    ///
+    /// Generated components are given incremental names (`source1`, `source2`,
+    /// etc) which should be replaced in order to provide better context.
+    expression: String,
+}
+
+#[derive(Serialize)]
+pub struct SinkOuter {
+    pub healthcheck: bool,
+    pub inputs: Vec<String>,
+    #[serde(flatten)]
+    pub inner: Value,
+    pub buffer: crate::buffers::BufferConfig,
+}
+
+#[derive(Serialize)]
+pub struct TransformOuter {
+    pub inputs: Vec<String>,
+    #[serde(flatten)]
+    pub inner: Value,
+}
+
+#[derive(Serialize, Default)]
+pub struct Config {
+    #[serde(flatten)]
+    pub global: GlobalOptions,
+    pub sources: IndexMap<String, Value>,
+    pub transforms: IndexMap<String, TransformOuter>,
+    pub sinks: IndexMap<String, SinkOuter>,
+}
+
+pub fn cmd(opts: &Opts) -> exitcode::ExitCode {
+    let components: Vec<Vec<String>> = opts
+        .expression
+        .split('|')
+        .map(|s| {
+            s.to_owned()
+                .split(',')
+                .map(|s| s.trim().to_owned())
+                .filter(|s| s.len() > 0)
+                .collect()
+        })
+        .collect();
+
+    let mut config = Config::default();
+    config.global.data_dir = crate::topology::config::default_data_dir();
+
+    let mut errs = Vec::new();
+
+    let mut i = 0;
+    let mut source_names = Vec::new();
+    components
+        .get(0)
+        .unwrap_or(&Vec::new())
+        .iter()
+        .for_each(|c| {
+            i += 1;
+            let name = format!("source{}", i);
+            source_names.push(name.clone());
+            config.sources.insert(name, {
+                let mut d = SourceDescription::example(c)
+                    .map_err(|e| {
+                        match e {
+                            ExampleError::MissingExample => {}
+                            _ => errs.push(e.clone()),
+                        }
+                        e
+                    })
+                    .unwrap_or(Value::Table(BTreeMap::new()));
+                d.as_table_mut().map(|s| {
+                    s.insert("type".to_owned(), c.to_owned().into());
+                    s
+                });
+                d
+            });
```

Comment:
Lots of nesting going on here too. My shot at a more idiomatic version:

```rust
let example = match SourceDescription::example(c) {
    Ok(example) => example,
    Err(err) => {
        if err != ExampleError::MissingExample {
            errs.push(e);
        }
        Value::Table(BTreeMap::new())
    }
}
example.as_table_mut().expect("examples are always tables").insert("type".into(), c.into()); 
config.sources.insert(name, example);
```

Reply:
Same as above, thanks!